src/bun.js/node/node_cluster_binding.zig

```diff
@@ -0,0 +1,280 @@
+            const ex = arguments_[0];
+            Process__emitErrorEvent(globalThis_, ex);
+            return .undefined;
+        }
+    };
+
+    const good = ipc_instance.data.serializeAndSendInternal(globalThis, message);
+
+    if (!good) {
+        const ex = globalThis.createTypeErrorInstance("sendInternal() failed", .{});
+        ex.put(globalThis, ZigString.static("syscall"), ZigString.static("write").toJS(globalThis));
+        const fnvalue = JSC.JSFunction.create(globalThis, "", S.impl, 1, .{});
+        Bun__Process__queueNextTick1(globalThis, fnvalue, ex);
+        return .false;
+    }
+
+    return .true;
+}
+
+pub fn onInternalMessageChild(globalThis: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(JSC.conv) JSC.JSValue {
+    log("onInternalMessageChild", .{});
+    const arguments = callframe.arguments(2).ptr;
+    child_singleton.worker = JSC.Strong.create(arguments[0], globalThis);
+    child_singleton.cb = JSC.Strong.create(arguments[1], globalThis);
+    child_singleton.flush(globalThis);
+    return .undefined;
+}
+
+pub fn handleInternalMessageChild(globalThis: *JSC.JSGlobalObject, message: JSC.JSValue) void {
+    log("handleInternalMessageChild", .{});
+
+    child_singleton.dispatch(message, globalThis);
+}
+
+//
+//
+//
+
+/// Queue for messages sent between parent and child processes in an IPC environment. node:cluster sends json serialized messages
+/// to describe different events it performs. It will send a message with an incrementing sequence number and then call a callback
+/// when a message is recieved with an 'ack' property of the same sequence number.
+pub const InternalMsgHolder = struct {
+    seq: i32 = 0,
+    callbacks: std.AutoArrayHashMapUnmanaged(i32, JSC.Strong) = .{},
+
+    worker: JSC.Strong = .{},
+    cb: JSC.Strong = .{},
+    messages: std.ArrayListUnmanaged(JSC.Strong) = .{},
+
+    pub fn isReady(this: *InternalMsgHolder) bool {
+        return this.worker.has() and this.cb.has();
+    }
+
+    pub fn enqueue(this: *InternalMsgHolder, message: JSC.JSValue, globalThis: *JSC.JSGlobalObject) void {
+        //TODO: .addOne is workaround for .append causing crash/ dependency loop in zig compiler
+        const new_item_ptr = this.messages.addOne(bun.default_allocator) catch bun.outOfMemory();
+        new_item_ptr.* = JSC.Strong.create(message, globalThis);
+    }
+
+    pub fn dispatch(this: *InternalMsgHolder, message: JSC.JSValue, globalThis: *JSC.JSGlobalObject) void {
+        if (!this.isReady()) {
+            this.enqueue(message, globalThis);
+            return;
+        }
+        this.dispatchUnsafe(message, globalThis);
+    }
+
+    fn dispatchUnsafe(this: *InternalMsgHolder, message: JSC.JSValue, globalThis: *JSC.JSGlobalObject) void {
+        const cb = this.cb.get().?;
+        const worker = this.worker.get().?;
+
+        const event_loop = globalThis.bunVM().eventLoop();
+
+        if (message.get(globalThis, "ack")) |p| {
```

Comment:
this still seems not great, perf-wise.

these messages should be a binary encoding, two integers. all the callbacks should be passed in ahead of time and have no dynamic arguments, and kept alive by the module loader continuing to be alive. if we need to keep the individual worker instances alive then that would be the only necessary strong value to hold, and we only need one of them in total.